src/java.base/share/classes/java/lang/runtime/SwitchBootstraps.java

```diff
@@ -235,13 +245,51 @@ public static CallSite enumSwitch(MethodHandles.Lookup lookup,
         labels = labels.clone();
 
         Class<?> enumClass = invocationType.parameterType(0);
+        Stream.of(labels).forEach(l -> validateEnumLabel(enumClass, l));
+        MutableCallSite result =
+                new MutableCallSite(invocationType);
+
+        MethodHandle temporary =
+                MethodHandles.insertArguments(LAZY_DO_ENUM_SWITCH, 2, labels, lookup, enumClass, result);
+        temporary = temporary.asType(invocationType);
+
+        result.setTarget(temporary);
+
+        return result;
+    }
+
+    private static <E extends Enum<E>> void validateEnumLabel(Class<?> enumClassTemplate, Object label) {
+        if (label == null) {
+            throw new IllegalArgumentException("null label found");
+        }
+        Class<?> labelClass = label.getClass();
+        if (labelClass == Class.class) {
+            if (label != enumClassTemplate) {
+                throw new IllegalArgumentException("the Class label: " + label +
+                                                   ", expected the provided enum class: " + enumClassTemplate);
+            }
+        } else if (labelClass == String.class) {
+            //OK
+        } else {
+            throw new IllegalArgumentException("label with illegal type found: " + labelClass +
+                                               ", expected label of type either String or Class");
+        }
+    }
+
+    private static int lazyDoEnumSwitch(Enum<?> target, int startIndex, Object[] labels, MethodHandles.Lookup lookup, Class<?> enumClass, MutableCallSite callSite) throws Throwable {
```

Comment:
can `doEnumSwitch` be folded into `lazyDoEnumSwitch`? just a suggestion, I'm OK with either way just that now it is not clear that we need two methods here. Also in `doEnumSwitch` and out of curiosity what example of user code could hit this section:

```
            if (label instanceof Class<?> c) {
                if (c.isAssignableFrom(targetClass))
                    return i;
            }
```

Reply:
The intent here is to avoid eager initialization of enums - so, until the enum is initialized, `lazyDoEnumSwitch` will be used (which, by itself, won't compute any results, except the result for `null`), which is then replaced with `doEnumSwitch` using the enum constants directly.

The enum switches support patterns as well, so something like:
```
enum E {A;}
E e = ...;
switch (e) {
     case E ee -> {}
}
```
is valid, and should trigger the part with Class labels.

Reply:
yep I found a similar test case triggering this code

Reply:
So, when the `enumSwitch` bootstrap is called, the enum class may or may not be initialized. So, to avoid initialization, we only do lightweight checks, and construct a `MutableCallSite` with `lazyDoEnumSwitch` as a target.

When the `CallSite` is invoked with an actual selector value, two options may arise:
a) the selector value is null. The `lazyDoEnumSwitch` itself returns -1.
b) the selector value is not null. The enum class must already be initialized (I believe). So, we create (hopefully) better structure to do the categorization, based on `doEnumSwitch`, and set that into the `CallSite`. Further calls will use this new `doEnumSwitch` based categorization directly, without calling `lazyDoEnumSwitch`. The current call will manually delegate to the `doEnumSwitch` version, but that should only happen "once" (or a small number of times, due to running in different threads, etc.).

Reply:
I see, thanks for the explanation, this is nice